[EuiDataGrid] improve height calculation

chandlerprall opened the pull request:
### Summary

Fixes a bug that was uncovered by adding the row height switcher, where going from single or line heights to auto fit wouldn't trigger a resize of the grid, as it is unable to guess the heights of rows when _auto_. This expands the unconstrained height calculation to include known heights and continue guessing at unknown rows, as auto-fit cells render their heights become known, causing the container to resize.

Added a cypress test for regression. Not sure if the approach grab the original height and do the rest of the test in a `.then()` block is the recommended approach, but it works and I couldn't find a way to do the comparison at a single block level.

### Checklist

~- [ ] Check against **all themes** for compatibility in both light and dark modes~
~- [ ] Checked in **mobile**~
- [x] Checked in **Chrome**, **Safari**, **Edge**, and **Firefox**
~- [ ] Props have proper **autodocs** and **[playground toggles](https://github.com/elastic/eui/blob/main/wiki/documentation-guidelines.md#adding-playground-toggles)**~
~- [ ] Added **[documentation](https://github.com/elastic/eui/blob/main/wiki/documentation-guidelines.md)**~
~- [ ] Checked **[Code Sandbox](https://codesandbox.io/)** works for any docs examples~
- [x] Added or updated **[jest](https://github.com/elastic/eui/blob/main/wiki/testing.md) and [cypress](https://github.com/elastic/eui/blob/main/wiki/cypress-testing.md) tests**
- [x] Checked for **breaking changes** and labeled appropriately
~- [ ] Checked for **accessibility** including keyboard-only and screenreader modes~
~- [ ] A **[changelog](https://github.com/elastic/eui/blob/main/wiki/documentation-guidelines.md#changelog)** entry exists and is marked appropriately~

chandlerprall commented on src/components/datagrid/body/data_grid_body.tsx:
```diff
@@ -573,11 +573,33 @@ export const EuiDataGridBody: FunctionComponent<EuiDataGridBodyProps> = (
+      knownRowCount++;
+      knownHeight += rowHeightUtils.getCalculatedHeight(
+        rowHeightOption,
+        defaultHeight, // minRowHeight,
```
Other calls to getCalculatedHeight have passed `minRowHeight` here, but `defaultHeight` looks like the better choice to me.

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_5447/

constancecchen commented on src/components/datagrid/body/data_grid_body.tsx:
```diff
@@ -573,11 +573,33 @@ export const EuiDataGridBody: FunctionComponent<EuiDataGridBodyProps> = (
     }
   }, [getRowHeight]);
 
+  let knownHeight = 0;
+  let knownRowCount = 0;
+  for (let i = startRow; i < endRow; i++) {
```
Should these calculations be occuring on every single rerender? I'd push for wrapping `knownHeight`/`knownRowCount` in either ` `useMemo` or `useState`/`useEffect`

chandlerprall replied:
I measured this and forgot to mention it in the description, thanks for asking! I called `performance.now()` before & after this snippet, and the difference was never more 2/10ths of a millisecond.

constancecchen replied:
Gotcha, thanks for checking!

Another reason I mentioned encapsulating `use` hooks around the new logic was not just for performance, but for logic organization. This isn't a blocker for this PR but might be worth considering for the future.

EuiDataGridBody is currently 500+ lines of disjointed calculations & effects ranging from rows to height to widths and frankly, it's something of a nightmare to parse as a new developer either trying to identify a specific problem or even just trying to read through the entire component to understand what all it does.

For example, without coming back to this PR and the specific bug/issue it's trying to resolve, I have no context for what `knownHeight` or `knownRowCount` is for or that it primarily applies to auto height grids. I'd like to see that documented more clearly in comments if possible.

constancecchen commented on src/components/datagrid/data_grid.spec.tsx:
```diff
@@ -51,6 +51,35 @@ describe('EuiDataGrid', () => {
     });
   });
+
+  describe('height calculation', async () => {
+    it('computes a new unconstrained height when switching to auto height', () => {
```
Huge ❤️ for the Cypress test!

> Not sure if the approach grab the original height and do the rest of the test in a .then() block is the recommended approach, but it works and I couldn't find a way to do the comparison at a single block level.

Yes, `.then()` chaining is generally the recommend Cypress approach IIRC / what I've also done in the past.

constancecchen commented:
Also for some reason my local Firefox for some reason just has never shown grid heights correctly (see ) so you might want someone else to corroborate that it works there

constancecchen commented:
Tested on Chrom, Edge, and Safari. Looks good except for one scenario:

- Set the localStorage example ([2nd on page](https://eui.elastic.co/pr_5447/#/tabular-content/data-grid-styling-and-control#adjusting-your-grid-to-usertoolbar-changes)) to `Auto fit`
- Refresh the page
- Note that the grid loads in with a scrollbar (which goes away when you click the display selector weirdly enough)

<img width="1159" alt="" src="https://user-images.githubusercontent.com/549407/144928537-05365cee-6abe-4a17-a1ff-88a1190e3bb4.png">

Also for some reason my local Firefox for some reason just has never shown grid heights correctly (see https://github.com/elastic/eui/pull/5313#pullrequestreview-789883367) so you might want someone else to corroborate that it works there

chandlerprall commented:
> Looks good except for one scenario:
> 
> Set the localStorage example (2nd on page) to Auto fit
> Refresh the page
> Note that the grid loads in with a scrollbar (which goes away when you click the display selector weirdly enough)

Pushed a fix, which triggers the grid body to re-render when any rows report a new height. If it looks good, I'll abstract the useState usage out so it's not such a weird block of code in the middle of other logic

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_5447/

constancecchen commented on src/components/datagrid/row_height_utils.ts:
```diff
@@ -173,6 +175,15 @@ export class RowHeightUtils {
     return Math.max(...rowHeightValues);
   }
 
+  onUpdatedRowHeight(callback: () => void) {
+    this.onUpdateCallbacks.push(callback);
+  }
+
+  offUpdatedRowHeight(callback: () => void) {
+    const idx = this.onUpdateCallbacks.indexOf(callback);
+    if (idx !== -1) this.onUpdateCallbacks.splice(idx, 1);
+  }
+
```
I can confirm this works with auto height and page refresh, but I'd strongly rather see this simplified. I don't see the array as necessary unless you anticipate more arbitrary callbacks in the future. I played around with this a bit locally and have a proposal for simplifying:

https://github.com/chandlerprall/eui/compare/bug/datagrid-height-calculation...constancecchen:bug/datagrid-height-calculation

If you're good with it, feel free to cherry-pick the commit! I strongly think `this.rerenderGridBody` is easier to grok/identify the purpose of at a glance than a generic `this.onUpdateCallbacks`, and would rather stick to a single fn than an array.

chandlerprall commented:
@constancecchen I've pushed three change sets:

* cherry-picked your callback simplification (thanks for that btw!)
* abstracted the force rerender callback into a hook
* refactored unconstrained height calculation into a hook for organization, added comments

constancecchen commented on src/services/hooks/useForceRender.ts:
```diff
@@ -0,0 +1,16 @@
+  const [, setRenderCount] = useState(0);
+  return useCallback(() => {
+    setRenderCount((x) => x + 1);
+  }, [setRenderCount]);
```
This is a super cool hook - love it!!

Optional comment - `setRenderCount` shouldn't ever change coming from setState, so you should be able to get rid of the dep

```suggestion
  }, []);
```

constancecchen commented on src/components/datagrid/body/data_grid_body.tsx:
```diff
@@ -254,6 +255,67 @@ export function getParentCellContent(_element: Node | HTMLElement) {
 }
 
+// computes the unconstrained (total possible) height of a grid
+const useUnconstrainedHeight = ({
```
This entire commit + set of comments is amazing and really helps me understand what this PR is doing. Thanks for taking the time to do it!! 😍 

In the future, I might poke a bit more at splitting up even more of `data_grid_body`'s logic into separate utils/hooks like this for organization - I personally think it's super useful for dev experience and unit testing. Thanks for the awesome headstart!

chandlerprall replied:
Breaking out into more hooks would be great; likely a great time to pair as well

constancecchen reviewed (approved):
This is seriously super cool. Thanks so much for going the extra mile here, I think it'll really pay off in the future in terms of dev experience and maintainability.

I have one minor non-blocking suggestion, but other than that I've tested this on Chrome, Safari, and Edge and confirm it's all working as expected. My Firefox is sadly still funkytown so if you could test that on my behalf it'd be super appreciated

chandlerprall commented:
In addition to Constance's 👁️ , I've tested in Chrome, FF, Edge, and Safari

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_5447/
